gpkit/keydict.py

```diff
@@ -0,0 +1,234 @@
+    particular `idx` parameter is used as a key.
+
+    By default a KeyDict will regenerate the list of possible key strings
+    for every usage; a KeyDict may instead be "baked" to have a fixed list of
+    keystrings by calling the `bake()` method.
+
+    Note that if a item is set using a key that does not have a `.key`
+    attribute, that key can be set and accessed normally.
+
+    See also: gpkit/tests/t_keydict.py.
+    """
+    collapse_arrays = True
+
+    def __init__(self, *args, **kwargs):
+        "Passes through to dict.__init__ via the `update()` method"
+        # pylint: disable=super-init-not-called
+        self.baked_keystrs = None
+        self.update(*args, **kwargs)
+
+    def update(self, *args, **kwargs):
+        "Iterates through the dictionary created by args and kwargs"
+        for k, v in dict(*args, **kwargs).items():
+            self[k] = v
+
+    def bake(self):
+        "Permanently sets the keystrs of a KeyDict"
+        self.baked_keystrs = None
+        self.baked_keystrs = self.keystrs()
+
+    def keystrs(self):
+        "Generates the strings that may be used as keys for a KeyDict"
+        if self.baked_keystrs:
+            return self.baked_keystrs
+        keystrs = defaultdict(set)
+        for key in self.keys():
+            if hasattr(key, "allstrs"):
+                for keystr in key.allstrs:
+                    keystrs[keystr].add(key)
+        return keystrs
+
+    @classmethod
+    def with_keys(cls, keyset, dictionaries):
+        "Generates a KeyDict from a KeySet and iterable of dictionaries"
+        out = cls()
+        for dictionary in dictionaries:
+            for key, value in dictionary.items():
+                # The keyset filters and converts each dictionary's keys
+                keys = keyset[key]
+                for key in keys:
+                    if not key.idx:
+                        out[key] = value
+                    else:
+                        if not hasattr(value, "shape"):
+                            value = np.array(value)
+                        val_i = value[key.idx]
+                        if is_sweepvar(val_i) or not np.isnan(val_i):
+                            out[key] = val_i
+        return out
+
+    def __contains__(self, key):
+        "In a winding way, figures out if a key is in the KeyDict"
+        if dict.__contains__(self, key):
+            return True
+        elif hasattr(key, "key"):
+            if dict.__contains__(self, key.key):
+                return True
+            elif self.is_veckey_but_not_collapsed(key):
+                if any(k in self for k in self.getkeys(key)):
+                    return True
+            elif self.is_index_into_vector(key.key):
+                vk = veckeyed(key.key)
+                if dict.__contains__(self, vk):
+                    return not bool(np.isnan(self.__dgi(vk)[key.key.idx]))
+        elif key in self.keystrs():
+            return True
```

Comment:
side note not related to the profiling effort -- looks to me like it's possible for this method to exit returning None -- shouldn't it `return False` at the very end?